src/Uno.UI/UI/Xaml/ResourceBindingCollection.cs

```diff
@@ -9,47 +9,69 @@
 using Uno.UI.DataBinding;
 using Windows.UI.Xaml.Data;
 
-namespace Windows.UI.Xaml
+namespace Windows.UI.Xaml;
+
+internal class ResourceBindingCollection
 {
-	internal class ResourceBindingCollection
-	{
-		private readonly Dictionary<DependencyProperty, Dictionary<DependencyPropertyValuePrecedences, ResourceBinding>> _bindings = new Dictionary<DependencyProperty, Dictionary<DependencyPropertyValuePrecedences, ResourceBinding>>();
+	private readonly Dictionary<DependencyProperty, Dictionary<DependencyPropertyValuePrecedences, ResourceBinding>> _bindings = new();
+	private BindingEntry[]? _cachedAllBindings;
+
+	public bool HasBindings => _bindings.Count > 0 && _bindings.Any(b => b.Value.Any());
 
-		public bool HasBindings => _bindings.Count > 0 && _bindings.Any(b => b.Value.Any());
+	public record struct BindingEntry(DependencyProperty Property, ResourceBinding Binding);
 
-		public IEnumerable<(DependencyProperty Property, ResourceBinding Binding)> GetAllBindings()
+	public BindingEntry[] GetAllBindings()
+	{
+		if (_cachedAllBindings is null)
 		{
+			List<BindingEntry> allBindings = new();
+
 			foreach (var kvp in _bindings)
 			{
 				foreach (var kvpInner in kvp.Value)
 				{
-					yield return (kvp.Key, kvpInner.Value);
+					allBindings.Add(new(kvp.Key, kvpInner.Value));
 				}
 			}
+
+			// We return a fully materialized list every time
+			// as the callers may enumerate the list and new items
+			// can be added when resource bindings are evaluated.
+			_cachedAllBindings = allBindings.ToArray();
 		}
 
-		public IEnumerable<ResourceBinding>? GetBindingsForProperty(DependencyProperty property)
-		{
-			if (_bindings.TryGetValue(property, out var bindingsForProperty))
-			{
-				return bindingsForProperty.Values;
-			}
+		return _cachedAllBindings;
+	}
 
-			return null;
+	public IEnumerable<ResourceBinding>? GetBindingsForProperty(DependencyProperty property)
+	{
+		if (_bindings.TryGetValue(property, out var bindingsForProperty))
+		{
+			return bindingsForProperty.Values;
 		}
 
-		public void Add(DependencyProperty property, ResourceBinding resourceBinding)
+		return null;
```

Comment:
Would not `Enumerable.Empty<ResourceBinding>()` (or the new C# 8.0 collection initializer) be handy here to avoid returning null when the method is expected to return a collection?

Reply:
In this case, returning null is cheaper than returning a empty emumerable, because it requires calling `IEnumerable<>.MoveNext()`, and we're in a critical path.

Reply:
Thanks